server/context.ts

```diff
@@ -1,28 +1,209 @@
     return createEmptyContext();
   }
 
+  const userAgent = req.headers.get("user-agent") || "";
+  const deviceNonce = req.headers.get("x-device-nonce") || "";
+  let ip = getClientIp(req);
+
+  if (process.env.NODE_ENV === "development") {
+    const ipInfo = await getIpInfo();
+    if (ipInfo) {
+      ({ ip } = ipInfo);
+    }
+  }
+
+  // Prioritize the custom auth token if present
+  if (customAuthHeader) {
+    console.log("Processing custom auth token");
+    try {
+      // Parse the base64 encoded JSON token using Node.js Buffer
+      let decodedTokenString;
+      try {
+        decodedTokenString = nodeDecode(customAuthHeader);
+        console.log("Decoded token string length:", decodedTokenString.length);
+      } catch (decodeError) {
+        console.error("Failed to decode custom auth token:", decodeError);
+        console.log("Raw token (first 50 chars):", customAuthHeader.substring(0, 50));
+        return createEmptyContext();
+      }
+      
+      let decodedToken;
+      try {
+        decodedToken = JSON.parse(decodedTokenString);
+        console.log("Parsed token with fields:", Object.keys(decodedToken));
+      } catch (parseError) {
+        console.error("Failed to parse token JSON:", parseError);
+        console.log("Decoded string:", decodedTokenString);
+        return createEmptyContext();
+      }
+      
+      // Validate the token (basic checks)
+      if (!decodedToken.user_id || !decodedToken.session_id) {
+        console.error("Invalid custom auth token format - missing required fields");
+        console.log("Token fields:", Object.keys(decodedToken));
+        return createEmptyContext();
+      }
+      
+      // Check token expiration
+      if (decodedToken.exp && decodedToken.exp < Math.floor(Date.now() / 1000)) {
+        console.error("Custom auth token expired");
+        return createEmptyContext();
+      }
+      
+      // Fetch the user from the database
+      const userProfile = await prisma.userProfile.findUnique({
+        where: { supId: decodedToken.user_id }
+      });
+      
+      if (!userProfile) {
+        console.error("User not found for custom auth token");
+        return createEmptyContext();
+      }
+      
+      console.log("Found user profile for custom auth:", {
+        id: userProfile.supId,
+        email: userProfile.supEmail ? "present" : "missing"
+      });
+      
+      // Look for the session in the database to validate
+      const sessionRecord = await prisma.session.findUnique({
+        where: { id: decodedToken.session_id }
+      });
+      
+      if (!sessionRecord) {
+        console.log("Session not found in database, creating new session record");
+        
+        // If session doesn't exist, create it based on the token data
+        const newSessionId = decodedToken.session_id || crypto.randomUUID();
+        const newDeviceNonce = decodedToken.device_nonce || deviceNonce || crypto.randomUUID();
```

Comment:
Device nonces should also be generated on the client.

Reply:
removed crypto.randomUUID(); for server generation